src/md.rs

```diff
@@ -1,19 +1,249 @@
+        const ASCII_ALPHABETIC  = 0b0000_0000_0000_0000_0000_0010_0000_0000;
+        const ASCII_GRAPHIC     = 0b0000_0000_0000_0000_0000_0100_0000_0000;
+        const ASCII_DIGIT       = 0b0000_0000_0000_0000_0000_1000_0000_0000;
+        const LATIN             = 0b0000_0000_0000_0000_0001_0000_0000_0000;
+        const ALPHABETIC        = 0b0000_0000_0000_0000_0010_0000_0000_0000;
+        const ACCENTUATED       = 0b0000_0000_0000_0000_0100_0000_0000_0000;
+        const CJK               = 0b0000_0000_0000_0000_1000_0000_0000_0000;
+        const HANGUL            = 0b0000_0000_0000_0001_0000_0000_0000_0000;
+        const KATAKANA          = 0b0000_0000_0000_0010_0000_0000_0000_0000;
+        const HIRAGANA          = 0b0000_0000_0000_0100_0000_0000_0000_0000;
+        const THAI              = 0b0000_0000_0000_1000_0000_0000_0000_0000;
+        const CASE_VARIABLE     = 0b0000_0000_0001_0000_0000_0000_0000_0000;
+        const LOWERCASE         = 0b0000_0000_0010_0000_0000_0000_0000_0000;
+        const UPPERCASE         = 0b0000_0000_0100_0000_0000_0000_0000_0000;
+        const NUMERIC           = 0b0000_0000_1000_0000_0000_0000_0000_0000;
+    }
+}
+
+#[derive(Copy, Clone)]
+pub struct MessDetectorChar {
+    pub character: char,
+    pub flags: MessDetectorCharFlags,
+    pub unicode_range: Option<&'static str>,
+}
+
+impl PartialEq for MessDetectorChar {
+    fn eq(&self, other: &Self) -> bool {
+        self.character == other.character
+    }
+}
+
+impl MessDetectorChar {
+    pub fn new(character: char) -> Self {
+        new_mess_detector_character(character)
+    }
+    fn in_category(
+        category: &str,
+        range: Option<&str>,
+        categories_exact: &[&str],
+        categories_partial: &[&str],
+        ranges_partial: &[&str],
+    ) -> bool {
+        // unicode category part
+        if categories_exact.contains(&category)
+            || categories_partial.iter().any(|&cp| category.contains(cp))
+        {
+            return true;
+        }
+        // unicode range part
+        if !ranges_partial.is_empty() {
+            if let Some(range) = range {
+                return ranges_partial.iter().any(|&r| range.contains(r));
+            }
+        }
+        false
+    }
+
+    fn in_description(name: Option<Name>, patterns: &[&str]) -> bool {
+        name.is_some_and(|description| {
+            patterns
+                .iter()
+                .any(|&s| description.to_string().contains(s))
+        })
+    }
+
+    pub fn is(&self, flag: MessDetectorCharFlags) -> bool {
+        self.flags.contains(flag)
+    }
+}
+
+#[cached(
+    type = "UnboundCache<char, MessDetectorChar>",
+    create = "{ UnboundCache::with_capacity(*UTF8_MAXIMAL_ALLOCATION) }",
+    convert = r#"{ character }"#
+)]
+pub fn new_mess_detector_character(character: char) -> MessDetectorChar {
+    let mut flags = MessDetectorCharFlags::empty();
+
+    // PLEASE NOTE! In case of idiomatic refactoring
```

Comment:
I think at least some of checkes down the line could be folded into one while keeping performance.

But maybe we could first merge if (tests pass / no performance regressions) and then worry about that later.

Reply:
Maybe you're right and match construction will be optimized by Rust, we'll see.

Reply:
Don't let that block current efforts though. We can always experiment with that later.